[lavanya] Added lavanya Telugu keyboard

prashanth-lavinna opened the pull request:
## Keyboard Name
Lavanya – Telugu Transliteration Keyboard

## Author
Prashanth Lavinna

## Summary
This keyboard is inspired by Lekhini and designed to allow Telugu typing using phonetic English input. It replicates Lekhini’s behavior closely with minor customizations. Anyone already familiar with Lekhini can use Telugu typing across any application, not just in a browser.

keyman-server commented:
Thank you for your pull request. You'll see a "build failed" message until the Keyman team has reviewed the pull request and manually initiated the build process.

Every change committed to this branch will become part of this pull request. When you have finished submitting files and are ready for the Keyman team to review this pull request, please post a "Ready for review" comment.

prashanth-lavinna commented:
Ready for review

LornaSIL commented:
Thanks for your Pull Request (PR)! There are a number of issues that I'll include here before we can accept the keyboard. It may seem like a lot, but I can see that you have basic issues okay, so you shouldn't have problems making these changes.

- In the .kmn file **Touch Layout**, the keycaps all have Latin characters. It does produce the right characters based on the desktop layout. I suspect users would want the keycaps to display Telugu?
- The welcome.htm file doesn't include any information on how to use the keyboard, but the `documentation.htm` file does. I would delete your current welcome.htm file, move `documentation.htm` into the `source` folder and rename it `welcome.htm`. 
   - I'm not seeing that any of your files reference the 3 graphics. Either remove the graphics or incorporate them into the new welcome.htm file.
- In the .kps file
   -  Two fonts are included:
      - One of them (Nirmala UI) is a Microsoft font and we cannot distribute that font as the license does not allow it. 
      - I found the other font `vani.ttf` online and it doesn't seem to have any license in it, so we cannot distribute that font either. 
      - I would recommend you either use the **Suranna UI** font that we include in the `release\shared\fonts\telu\suranna_ui` or we could add the Noto Sans Telugu or Noto Serif Telugu font to our shared\fonts repo. Let me know which you prefer to use. Whatever font gets packaged must be under an Open License such as OFL, MIT, GNU, etc. If you want to add one of the Noto Telugu fonts to  `release\shared\fonts\noto\telu` you may do so, or I can do it first before you make your changes. Just let me know your preference.
      - The alternative is to not include a font in the package and the user will just get whatever comes on their system. Since all operating systems will support the Telugu script that's not a bad idea either.
   - Now, also in the .kps file you need to include your new `welcome.htm` in the list of files. If you decided to include any graphics in that file, you should add those to the Files to include.
   - On the **Keyboards** tab, if you added one of the fonts I suggested, you should include it for the **Keyboard font** and for the **Display font**.
   - Under the **Details** tab you should add welcome.htm.
      - For the Description field I would add more information such as "This custom Telugu keyboard is inspired by Lekhini, a browser-based transliteration tool. It was created to overcome Lekhini’s limitation of working only within a browser. Most of the character mappings from Lekhini have been faithfully replicated, with a few exceptions."
- The `lavanya.php` file needs some changes to display properly online. Remove the first 83 lines and replace it with:
```
<?php
  $pagename = 'Lavanya Keyboard Help';
    $pagetitle = $pagename;
    $pagestyle = <<<END
        body { font-family: Nirmala UI, Vani, 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;  }
        table { border-collapse: collapse; width: 100%; margin: 20px 0; }
        th { background-color: #e8eaf6; padding: 10px; text-align: left;  border: 1px solid #c5cae9; }
        td { padding: 8px 10px; border: 1px solid #c5cae9; }
        tr:nth-child(even) { background-color: #f5f5f5; }
        code { background-color: #f5f5f5; padding: 2px 4px; border-radius: 3px; font-family: 'Courier New', Courier, monospace; }
        .example { background-color: #e3f2fd; padding: 15px; border-left: 4px solid #2196f3; margin: 20px 0; }
        ul, ol { margin-left: 20px; }
        li { margin-bottom: 8px; }
  END;
  require_once('header.php');
?>
```
- Then, because php should include the body and html tags, and because our online system handles the footer, you should delete these last lines in the file:
```
    <footer>
        <p>Last updated: <?php echo $currentDate; ?></p>
        <p>Generated by: <?php echo $_SERVER['SERVER_NAME']; ?></p>
    </footer>
</body>
</html>
```

I think those are most of the changes. I'll have to review it again after you make the changes.

prashanth-lavinna commented:
@LornaSIL 
I've made the following changes as requested. 
- I've updated the 'Touch Layout' to include Telugu Characters
- I copied content from documentation file to welcome file. Deleted the documentation file. Welcome file is at /source/welcome/welcome.htm. I also included welcome file in the list of files in package file. 
- I utilized 2 of the 3 graphic files. Deleted the 3rd one. 
- I included 'Noto Sans Telugu' font in files. The name of the font I downloaded was 'NotoSansTelugu-VariableFont_wdth,wght'. I didn't change this name. Hoping that's not issue. Chose the same font in both Keyboard Font and Display Font
- I added welcome file in the details tab. 
- Updated the .php file. 

*I apologize in advance if I still committed any mistakes. This is my first time handling a Git Hub Pull Request.*

LornaSIL commented:
I think the main remaining issue is that the font is in your downloads folder and not in the github repo. It should be added to `release\shared\fonts\noto\telu` Then, you'll need to remove it from the list of files and then add it back in from the new location. If you don't have the whole repo cloned, you'll need to fake the folder structure so it's equivalent to how the repo is structured. If you don't have the repo cloned it might be easier for me to add it and then merge master into your PR.

Also, this will be an experiment because I don't think we have any other Variable fonts in the repo at this time. We'll try it and see if it works okay.

prashanth-lavinna commented:
@LornaSIL 
I deleted the existing font. Saved the font @ `release\shared\fonts\noto\Telu` and reuploaded in package source file. 

**A question for future:**
In my previous commit, I made the requested changes outside the repository. After building the package there, I deleted the existing files in my version-controlled branch and copied the updated files into the Keyman repository.

In this commit, I directly opened the `.kmn` and `.kps` files from my branch in the Keyman Developer application and made the requested changes within the repository.

From your perspective, which of these two approaches is preferred?

LornaSIL reviewed (approved):
I made some changes to your pull request. Please make sure to grab the latest files before you make any further updates to the keyboard.
Otherwise, it looks good and I'll merge it in! Congrats!

LornaSIL commented:
Your keyboard is now online: https://keyman.com/keyboards/lavanya
The online help file has not deployed yet, but hopefully soon.

DavidLRowe commented:
> **A question for future:** In my previous commit, I made the requested changes outside the repository. After building the package there, I deleted the existing files in my version-controlled branch and copied the updated files into the Keyman repository.
> 
> In this commit, I directly opened the `.kmn` and `.kps` files from my branch in the Keyman Developer application and made the requested changes within the repository.
> 
> From your perspective, which of these two approaches is preferred?

Whatever is easiest for you! Both methods should work. When copying files, you'll need to make sure they end up in the correct location.

DavidLRowe commented:
The https://help.keyman.com/knowledge-base/?id=106 article shows one way to download the updated keyboard folder to make sure you have the exact files that are in the repository.

prashanth-lavinna commented:
Thank you David Rowe and Lorna Evans. I appreciate you and your teams
efforr a lot. Long live open source.

Prashanth

On Fri, May 9, 2025, 4:29 PM DavidLRowe ***@***.***> wrote:

> *DavidLRowe* left a comment (keymanapp/keyboards#3502)
> <https://github.com/keymanapp/keyboards/pull/3502#issuecomment-2867809462>
>
> The https://help.keyman.com/knowledge-base/?id=106 article shows one way
> to download the updated keyboard folder to make sure you have the exact
> files that are in the repository.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keymanapp/keyboards/pull/3502#issuecomment-2867809462>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BJFNDPUGSBTKX2SFAMEXG5D25UF25AVCNFSM6AAAAAB4VO44WGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNRXHAYDSNBWGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

LornaSIL commented:
You should open a new Pull Request, and when you make changes to the keyboard always "bump" (increase) the version number in the .kmn file and document the change in `HISTORY.md`.
